### PR TITLE
[serve] Gracefully stop replicas that fail health checks

### LINKED ISSUE #41742: [Serve] Gracefully shut down unhealthy replicas
### Description

Currently, when replicas fail enough health checks, they're considered unhealthy. Serve then shuts down these unhealthy replicas immediately.

Instead, Serve should shut them down gracefully.

### Use case

This gives unhealthy replicas extra time to finish processing requests.

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

See https://github.com/ray-project/ray/issues/41742 for context.

## Related issue number

Closes https://github.com/ray-project/ray/issues/41742

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/deployment_state.py [hunk 1]
```diff
@@ -1018,7 +1019,7 @@ def check_stopped(self) -> bool:
         if self._actor.check_stopped():
             return True
 
-        timeout_passed = time.time() > self._shutdown_deadline
+        timeout_passed = time.time() >= self._shutdown_deadline
```
Comment (edoakes) [inline python/ray/serve/_private/deployment_state.py:1022]: tiny edge case that I updated to make writing unit tests more convenient

**Review by shrekris-anyscale — APPROVED** [review]
Nice work, thanks for taking this on!
